libcst/metadata/base_provider.py

```diff
@@ -44,9 +44,17 @@ class BaseMetadataProvider(MetadataDependent, Generic[_ProvidedMetadataT]):
     # explanation.
     _computed: MutableMapping["CSTNode", _ProvidedMetadataT]
 
-    def __init__(self) -> None:
+    is_cache_required: bool = False
+
+    def __init__(self, cache: object = None) -> None:
```

Comment:
It's probably good enough for an MVP, but I'm not particularly happy about `cache` being typed as `object`. Is there a way we could make this a typevar? Could we infer `is_cache_required` from the generic type parameter?

Reply:
typing_inspect.get_parameters

Reply:
I would not use `get_parameters` to make caching depending on the existence of an init param. Instead, either we should use type generics as Ben suggested, or provide a cache-required base class and a no-cache-required base class, and check for `isinstance` of either to determine cache requirements.

Reply:
The second type generic requires changing the signature of existing metadata provider which will break our custom metadata provider in distillery.